public/gameInfo.php

```diff
@@ -234,20 +244,29 @@
         array_multisort($authorCount, SORT_DESC, $authorInfo);
     }
 
-    // If the game is beatable, the user has met the requirements to receive the
-    // beaten game award, and they do not currently have that award, give it to them.
+    // Show the beaten award display in the progress component optimistically.
+    // The actual award metadata is updated async via actions/background jobs.
     if ($isGameBeatable) {
         $neededProgressions = $totalProgressionAchievements > 0 ? $totalProgressionAchievements : 0;
         $neededWinConditions = $totalWinConditionAchievements > 0 ? 1 : 0;
-        if (
-            $totalEarnedProgression === $neededProgressions
+
+        $isBeatenSoftcore = (
+            $totalEarnedProgression === $totalProgressionAchievements
             && $totalEarnedWinCondition >= $neededWinConditions
-            && !$isBeatenHardcore
-            && !$isBeatenSoftcore
-        ) {
-            $beatenGameRetVal = testBeatenGame($gameID, $user, true);
-            $isBeatenHardcore = $beatenGameRetVal['isBeatenHardcore'];
-            $isBeatenSoftcore = $beatenGameRetVal['isBeatenSoftcore'];
+        );
+
+        $isBeatenHardcore = (
+            $totalEarnedProgressionHardcore === $totalProgressionAchievements
+            && $totalEarnedWinConditionHardcore >= $neededWinConditions
+        );
+
+        // TODO: Remove this side effect when switching to aggregate queries.
+        // Without aggregate queries, the side effect is part of the beaten games
+        // self-healing mechanism.
+        if (!config('feature.aggregate_queries')) {
```

Comment:
You could move this up to line 249. There's no reason to calculate `$isBeatenSoftcore` or `$isBeatenHardcore` unless it goes into this if.

Reply:
I'm not sure I follow - the optimistic calculation without the `testBeatenGame()` call is the purpose of the PR.

Previously, these values were being set from either the SiteAwards table or based on the return value of `testBeatenGame()`.

Based on [this conversation](https://github.com/RetroAchievements/RAWeb/pull/1874#discussion_r1335158595) in #1874, there is a desire to remove the `testBeatenGame()` side effect from the game page entirely.

However, due to the upcoming async nature of the award updates in the back-end, we can't guarantee the user won't erroneously see a status label for a beaten award they should or shouldn't have. By deriving the award status optimistically through the achievement data already loaded for the page, we can mitigate such scenarios.

Reply:
> I'm not sure I follow - the optimistic calculation without the testBeatenGame() call is the purpose of the PR.

Logic before:
```
        if (
            $totalEarnedProgression === $neededProgressions
            && $totalEarnedWinCondition >= $neededWinConditions
            && !$isBeatenHardcore
            && !$isBeatenSoftcore
        ) {
            $beatenGameRetVal = testBeatenGame($gameID, $user, true);
```

It doesn't look like the logic that gates the call to `testBeatenGame` has changed. The first two lines become `$isBeatenSoftcore`, and the last line is the logical comparison that the award hasn't been previously given.

New logic:
```
        $isBeatenSoftcore = (
            $totalEarnedProgression === $totalProgressionAchievements
            && $totalEarnedWinCondition >= $neededWinConditions
        );
        
        ...
        
        if ($isBeatenSoftcore !== $hasBeatenSoftcoreAward || $isBeatenHardcore !== $hasBeatenHardcoreAward) {
            $beatenGameRetVal = testBeatenGame($gameID, $user, true);
        }
````

Reply:
> However, if you then changed a 9th achievement to be progression, you would continue to see the label even though you hadn't earned that 9th progression achievement. After the changes in this PR, that is no longer the case.

I don't think that would have worked before either as this would not have been true: `$totalEarnedProgression === $neededProgressions`

Reply:
> I don't think that would have worked before either as this would not have been true:
`$totalEarnedProgression === $neededProgressions`

You're right, it didn't work before either. It does now with this change.

**Before**
In this screenshot, an achievement ("Pro Collector", last on the list in the image) has been given a Progression type _after_ I've earned the award. I still visually see the "Beaten" status text, even though I shouldn't.

![Screenshot 2023-09-30 at 7 34 25 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/ca5a8e38-1443-4324-ae9f-807ff533dae6)

**After**
Changing nothing else besides switching from `master` to this branch, the label now says "Unfinished", even though I technically still have the award in the database. Because of `$isBeatenHardcore` and `$isBeatenSoftcore` being optimistically adjusted, the erroneous label no longer displays. Eventually, the awards data in the database will catch up.

![Screenshot 2023-09-30 at 7 36 20 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/3379edf8-0201-43a2-9e5c-17cf1d702747)